src/string_radix_sort.jl

```diff
@@ -0,0 +1,386 @@
+import Base: Forward, ForwardOrdering, Reverse, ReverseOrdering, Lexicographic, LexicographicOrdering, sortperm, Ordering, 
+            setindex!, getindex, similar
+# import SortingAlgorithms: RadixSort, RadixSortAlg
+"""
+    load_bits([type,] s, skipbytes)
+
+Load the underlying bits of a string `s` into a `type` of the user's choosing.
+The default is `UInt`, so on a 64 bit machine it loads 64 bits (8 bytes) at a time.
+If the `String` is shorter than 8 bytes then it's padded with 0.
+
+- `type`:       any bits type that has `>>`, `<<`, and `&` operations defined
+- `s`:          a `String`
+- `skipbytes`:  how many bytes to skip e.g. load_bits("abc", 1) will load "bc" as bits
+"""
+ # Some part of the return result should be padded with 0s.
+# To prevent any possibility of segfault we load the bits using
+# successively smaller types
+# it is assumed that the type you are trying to load into needs padding
+# i.e. `remaining_bytes_to_load > 0`
+# load_bits_with_padding(::Type{UInt128}, s::String, skipbytes=0)= load_bits_with_padding(UInt128, s, skipbytes, [UInt64, UInt32, UInt16, UInt8])
+# load_bits_with_padding(::Type{UInt64}, s::String, skipbytes=0) = load_bits_with_padding(UInt64, s, skipbytes, [UInt32, UInt16, UInt8])
+# load_bits_with_padding(::Type{UInt32}, s::String, skipbytes=0) = load_bits_with_padding(UInt32, s, skipbytes, [UInt16, UInt8])
+
+# function load_bits_with_padding(::Type{T}, s::String, skipbytes, smaller_types) where T <: Unsigned    
+#     n = sizeof(s)
+
+#     ns = (sizeof(T) - min(sizeof(T), n - skipbytes))*8
+#     remaining_bytes_to_load = sizeof(T) - ns÷8
+#     # start
+#     res = zero(T)
+#     shift_for_padding = sizeof(T)*8
+
+#     smaller_sizes = sizeof.(smaller_types)
+#     for (S, type_size) in zip(smaller_types, smaller_sizes)
+#         if  remaining_bytes_to_load >= type_size
+#             res |= Base.zext_int(T, ntoh(unsafe_load(Ptr{S}(pointer(s, skipbytes+1))))) << (shift_for_padding - type_size*8)
+#             skipbytes += type_size
+#             remaining_bytes_to_load -= type_size
+#             shift_for_padding -= type_size*8
+#         end
+#     end
+#     return res
+# end
+
+function load_bits_with_padding(::Type{UInt128}, s::String, skipbytes = 0)  
+    n = sizeof(s)
+    T = UInt128
```

Comment:
I'm not sure the compiler is smart enough to constant-fold a type that you assign to a local variable like this.  Does it improve performance if you just paste `UInt128` directly in the code?

(I think it should be possible to write this function in a way that works for all `T` simultaneously with no penalty by making sure the branches get optimized out, but easier to optimize the case for a single `T` first.)

Reply:
Ok, I will just replace wiht
Actually I think a macro will work better here; just generate the code for all types. It's basically the same anyway. Might look into that.

I actually initial I have one function for all types. But it keeps on crashing. I think it has to do with the `Base.zext_int(UInt64, ::UInt128)` which doesn't make sense; even that branch will never be reached, the compiler still complained and crashed the program every time.